File: backend/app/schemas/firebase.py
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field, ConfigDict
from datetime import datetime
from enum import Enum


class PermissionLevel(int, Enum):
    VIEWER = 1
    COMMENTER = 2
    EDITOR = 3
    OWNER = 4


class AIFeature(str, Enum):
    SUMMARIZE = "summarize"
    TRANSLATE = "translate"
    GRAMMAR_CHECK = "grammar-check"
    TONE_ADJUSTMENT = "tone-adjustment"
    EXPAND = "expand"
    SHORTEN = "shorten"
    GENERATE_OUTLINE = "generate-outline"
    HELP_ME_WRITE = "help-me-write"
    IMPROVE = "improve"


class FirebaseDocumentCreate(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    title: str = "Untitled Document"
    owner_id: str = Field(alias="ownerId")


class FirebaseDocumentUpdate(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    title: Optional[str] = None
    ai_consent: Optional[bool] = Field(default=None, alias="aiConsent")
    ai_features: Optional[List[str]] = Field(default=None, alias="aiFeatures")


class FirebaseDocumentContent(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    content: Dict[str, Any]


class FirebaseDocument(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    title: str
    owner_id: str = Field(alias="ownerId")
    permissions: Dict[str, int]
    content_hash: str = Field(default="", alias="contentHash")
    storage_path: str = Field(alias="storagePath")
    created_at: datetime = Field(alias="createdAt")
    updated_at: datetime = Field(alias="updatedAt")
    version: int = 1
    ai_consent: bool = Field(default=False, alias="aiConsent")
    ai_features: List[str] = Field(default_factory=list, alias="aiFeatures")
    is_deleted: bool = Field(default=False, alias="isDeleted")
    deleted_at: Optional[datetime] = Field(default=None, alias="deletedAt")


class AIRequestCreate(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    user_id: str
    document_id: str
    feature: AIFeature
    input: Dict[str, Any] = Field(default_factory=dict)
    ai_consent: bool = Field(default=False, alias="aiConsent")


class AIRequestResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    user_id: str = Field(alias="userId")
    document_id: str = Field(alias="documentId")
    feature: AIFeature
    status: str
    input: Dict[str, Any]
    output: Optional[Dict[str, Any]] = None
    tokens_used: Optional[int] = Field(default=None, alias="tokensUsed")
    created_at: datetime = Field(alias="createdAt")
    completed_at: Optional[datetime] = Field(default=None, alias="completedAt")
    error: Optional[str] = None


class ShareInvitationCreate(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    document_id: str = Field(alias="documentId")
    owner_id: str = Field(alias="ownerId")
    invitee_email: str = Field(alias="inviteeEmail")
    permission_level: PermissionLevel = Field(alias="permissionLevel")


class ShareInvitation(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    document_id: str = Field(alias="documentId")
    inviter_id: str = Field(alias="inviterId")
    invitee_email: str = Field(alias="inviteeEmail")
    permission_level: int = Field(alias="permissionLevel")
    status: str
    created_at: datetime = Field(alias="createdAt")
    expires_at: datetime = Field(alias="expiresAt")

File: frontend/src/types/firebase.ts
export type PermissionLevel = 1 | 2 | 3 | 4

export const PERMISSIONS = {
  VIEWER: 1 as PermissionLevel,
  COMMENTER: 2 as PermissionLevel,
  EDITOR: 3 as PermissionLevel,
  OWNER: 4 as PermissionLevel,
}

export interface User {
  id: string
  email: string | null
  displayName: string | null
  photoURL: string | null
  createdAt: Date
  lastLoginAt: Date
  settings?: UserSettings
}

export interface UserSettings {
  theme: 'light' | 'dark' | 'system'
  fontSize: 'small' | 'medium' | 'large'
  aiConsent: boolean
  emailNotifications: boolean
}

export interface Document {
  id: string
  title: string
  ownerId: string
  permissions: Record<string, PermissionLevel>
  contentHash: string
  storagePath: string
  createdAt: Date
  updatedAt: Date
  version: number
  aiConsent: boolean
  aiFeatures: string[]
  isDeleted: boolean
  deletedAt?: Date
}

export interface DocumentContent {
  type: 'doc'
  content: Array<{
    type: string
    content?: Array<{
      type: string
      text?: string
      marks?: Array<{ type: string }>
      attrs?: Record<string, unknown>
    }>
    attrs?: Record<string, unknown>
  }>
}

export interface UserPresence {
  userId: string
  displayName: string
  photoURL: string | null
  color: string
  lastActive: Date
  cursorPosition?: {
    from: number
    to: number
  }
}

export type AIFeature =
  | 'summarize'
  | 'translate'
  | 'grammar-check'
  | 'tone-adjustment'
  | 'expand'
  | 'shorten'
  | 'generate-outline'
  | 'help-me-write'
  | 'improve'

export interface AIRequest {
  id: string
  userId: string
  documentId: string
  feature: AIFeature
  status: 'pending' | 'processing' | 'completed' | 'failed'
  input: Record<string, unknown>
  output?: Record<string, unknown>
  tokensUsed?: number
  createdAt: Date
  completedAt?: Date
  error?: string
}

export interface ShareInvitation {
  id: string
  documentId: string
  inviterId: string
  inviteeEmail: string
  permissionLevel: PermissionLevel
  status: 'pending' | 'accepted' | 'declined' | 'expired'
  createdAt: Date
  expiresAt: Date
}

export function assignCursorColor(index: number): string {
  const colors = [
    '#FF6B6B',
    '#4ECDC4',
    '#45B7D1',
    '#96CEB4',
    '#FFEAA7',
    '#DDA0DD',
    '#98D8C8',
    '#F7DC6F',
    '#BB8FCE',
    '#85C1E9',
  ]
  return colors[index % colors.length]
}
